lib/stm32/f4/rcc.c

```diff
@@ -446,60 +458,78 @@ void rcc_osc_bypass_disable(enum rcc_osc osc)
 	}
 }
 
+/*
+ */
+void rcc_set_sysclk(enum rcc_osc clk) {
+	uint32_t clk_bits = RCC_CFGR_SW_HSI;
+	switch (clk) {
+		case HSI:
+			clk_bits = RCC_CFGR_SW_HSI;
+			break;
+		case HSE:
+			clk_bits = RCC_CFGR_SW_HSE;
+			break;
+		case PLL:
+			clk_bits = RCC_CFGR_SW_PLL;
+			break;
+		default:
+			clk_bits = RCC_CFGR_SW_HSI;
+			break;
+	}
+	RCC_CFGR = (RCC_CFGR & (~RCC_CFGR_SW_MASK)) |
+			   (clk_bits & RCC_CFGR_SW_MASK);
+	/* wait for the switch */
+	while (((RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SW_MASK) != clk_bits) ;
+}
 
+enum rcc_osc rcc_get_sysclk(void) {
+	uint32_t clk_bits = (RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SW_MASK;
+	switch (clk_bits) {
+		default:
+			return HSI;
+		case 1:
+			return HSE;
+		case 2:
+			return PLL;
+		case 3:
+			return HSI; // Error not sure what to do here
+	}
+}
 
 void rcc_set_sysclk_source(uint32_t clk)
 {
-	uint32_t reg32;
-
-	reg32 = RCC_CFGR;
-	reg32 &= ~((1 << 1) | (1 << 0));
-	RCC_CFGR = (reg32 | clk);
+	RCC_CFGR = (RCC_CFGR & (~RCC_CFGR_SW_MASK)) |
+		   (clk & RCC_CFGR_SW_MASK);
 }
 
 void rcc_set_pll_source(uint32_t pllsrc)
 {
-	uint32_t reg32;
-
-	reg32 = RCC_PLLCFGR;
-	reg32 &= ~(1 << 22);
-	RCC_PLLCFGR = (reg32 | (pllsrc << 22));
+	RCC_PLLCFGR = (RCC_PLLCFGR & (~RCC_PLLCFGR_PLLSRC)) |
+			  ((pllsrc & 1) * RCC_PLLCFGR_PLLSRC);
 }
 
 void rcc_set_ppre2(uint32_t ppre2)
 {
-	uint32_t reg32;
-
-	reg32 = RCC_CFGR;
-	reg32 &= ~((1 << 13) | (1 << 14) | (1 << 15));
-	RCC_CFGR = (reg32 | (ppre2 << 13));
+	RCC_CFGR = (RCC_CFGR & (~(RCC_CFGR_PPRE_MASK << RCC_CFGR_PPRE2_SHIFT))) |
+		   ((ppre2 & RCC_CFGR_PPRE_MASK) << RCC_CFGR_PPRE2_SHIFT);
```

Comment:
Actually I think all of these register setting functions are candidates to be made into static inline functions inside the rcc.h file. Really they are just bit manipulation in a register.